src/libDirectoryService/ShardingPreProcessing.cpp

```diff
@@ -88,131 +88,82 @@ void DirectoryService::ComputeSharding()
-              m_mediator.m_selfPeer);
-    curr_offset += IP_SIZE + PORT_SIZE;
+    // For this version, DS leader (the one invoking this function) is part of the X nodes to receive and share Tx bodies -> fill up leader's assigned nodes
+
+    m_sharingAssignment.clear();
 
-    for (unsigned int i = m_consensusMyID + 1; i < num_ds_nodes; i++)
+    for (unsigned int i = num_ds_nodes; i < m_mediator.m_DSCommittee.size();
+         i++)
     {
-        m_mediator.m_DSCommittee.at(i).second.Serialize(sharding_structure,
-                                                        curr_offset);
-        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                  m_mediator.m_DSCommittee.at(i).second);
-        curr_offset += IP_SIZE + PORT_SIZE;
+        m_sharingAssignment.emplace_back(m_mediator.m_DSCommittee.at(i).second);
     }
 
-    LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-              "Number of shards: " << m_shards.size());
-
-    Serializable::SetNumber<uint32_t>(sharding_structure, curr_offset,
-                                      (uint32_t)m_shards.size(),
-                                      sizeof(uint32_t));
-    curr_offset += sizeof(uint32_t);
-
     // PART 2 and 3
     // First version: We just take the first X nodes for receiving and next X nodes for sending
+
+    m_shardReceivers.clear();
+    m_shardSenders.clear();
+
     for (unsigned int i = 0; i < m_shards.size(); i++)
     {
         const map<PubKey, Peer>& shard = m_shards.at(i);
 
-        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                  "Shard " << i << " forwarders:");
-
         // PART 2
+
+        m_shardReceivers.emplace_back(vector<Peer>());
+
         uint32_t nodes_recv_lo = 0;
         uint32_t nodes_recv_hi = nodes_recv_lo + TX_SHARING_CLUSTER_SIZE - 1;
+
         if (nodes_recv_hi >= shard.size())
         {
             nodes_recv_hi = shard.size() - 1;
         }
 
         unsigned int num_nodes = nodes_recv_hi - nodes_recv_lo + 1;
 
-        Serializable::SetNumber<uint32_t>(sharding_structure, curr_offset,
-                                          num_nodes, sizeof(uint32_t));
-        curr_offset += sizeof(uint32_t);
-
         map<PubKey, Peer>::const_iterator node_peer = shard.begin();
         for (unsigned int j = 0; j < num_nodes; j++)
         {
-            node_peer->second.Serialize(sharding_structure, curr_offset);
-            curr_offset += IP_SIZE + PORT_SIZE;
-
-            LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      node_peer->second);
-
+            m_shardReceivers.back().emplace_back(node_peer->second);
             node_peer++;
         }
 
-        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                  "Shard " << i << " senders:");
-
         // PART 3
+
+        m_shardSenders.emplace_back(vector<Peer>());
```

Comment:
The reason we recommend emplace_back is to reduce the object construction overhead. Thus it's better to just write m_shardReceivers.emplace_back() here.

Reply:
Done